[PVR] Allow edit of timers with start/end in the past

metaron-uk opened the pull request:
Allows Start/End to be a value earlier than now() when editing a timer
Turning off 'Any time' defaults to FirstDay time, rather than now() _(EDIT: This bit reverted pending separate API change PR for 'AnyTime'.)_
If start<end assume the timer crosses midnight and adjust as necessary

@ksooo - this resolves the first comment I had on https://github.com/xbmc/xbmc/pull/7508

Tested against my pvr.mythtv working version (https://github.com/metaron-uk/pvr.mythtv/tree/reducedEPG)

metaron-uk committed:
[PVR] Timer settings dialog: Start/end in the past

Allows Start/End to be a value earlier than now() when editing a timer
For timers with start/end clocks but no start/end day spinners:
If start<end
 - assume timer crosses midnight and add a day
 - if this still fails, set end=start
If end>start+1day
 - assume midnight was removed and remove a day
 - if this still fails, set end=start

Types with start/end date spinners just log a warning if start<end

ksooo commented:
@metaron-uk rearding problem 2 you mentioned in #7508: https://github.com/ksooo/xbmc/commit/ba44b3b0a68e5f515baa181d7491fbd288f1ec96 Does this work for you?

metaron-uk commented:
@ksooo While checking this fix out, I noticed that you can't actually set a first day in the past :( 
Seems it assumes you mean next year when you select down on the spinner. Moral: never do regression testing when really tired! 
I will investigate.
![screenshot004](https://cloud.githubusercontent.com/assets/12870817/8781403/add886a4-2f06-11e5-9154-f1596ca294b0.png)

**Also:**
strSummary seems to use firstDay - even when zero - to set the date of repeating timers which have subordinates, even if they have a valid start/end time. It uses start/end for for timers without subordinates!
It's even more confusing when you consider the sort order - this seems not to respect firstDay even when using it to populate strSummary
![screenshot002](https://cloud.githubusercontent.com/assets/12870817/8781375/81e9382c-2f06-11e5-9507-52f4ddb6fde5.png)
![screenshot003](https://cloud.githubusercontent.com/assets/12870817/8781376/81f14c74-2f06-11e5-9c84-4da4c6c53d6d.png)

All the timers before 'great barrier reef' have subordinates, the ones afterwards don't.
In these examples my pvr.mythtv code sets firstDay only for 'Record weekly' and 'Record daily' timer types. All the other types set start & end based on backend data (which is the start/end of the EPG entry used to create the timer in the 1st place), but not 1st day.
I will investigate this as well.

ksooo commented:
All this first day stuff makes me sick, especially as it is not used at all for tvheadend. Thus, you're almost on your own with this when it comes to testing. The only thing I'm really scared of is the high potential for regressions for other add-ons if you change things the mythtv way...

metaron-uk commented:
I understand.

The best solution to avoid issues imo is to keep first day & start / end time management completely separate - that way if you don't use one, it won't affect the others.
The important thing is that users can create and edit a timer without Kodi forcing times and dates to always occur in the future.
I'll re-work this PR to strip out the relationships between firstday/starttime/endtime and try to fix the remining issues with them being forced into the future (If you've any ideas about where to look for these please let me know).

Separately, there are two features which I would like PVR core to support (which it doesn't with the current API - I think we have tried hard enough to force it!):
1. Getting a meaningful 'time' to sort timers in 'date' order without breaking the underlying timer settings.
2. Allowing a user to toggle 'Any Time' without loosing the start and end time values.

I think this warrants a better explanation, but I'll move that to the forum.

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -116,8 +116,9 @@ void CGUIDialogPVRTimerSettings::SetTimer(CFileItem *item)
   const CDateTime now(CDateTime::GetCurrentDateTime());
-  m_startLocalTime    = m_timerInfoTag->IsStartAtAnyTime() ? now : InitializeDateTime(m_timerInfoTag->StartAsLocalTime());
-  m_endLocalTime      = m_timerInfoTag->IsEndAtAnyTime()   ? now : InitializeDateTime(m_timerInfoTag->EndAsLocalTime());
+  bool bFirstDay((m_timerType && m_timerType->SupportsFirstDay()) || (GetDateAsInt(m_timerInfoTag->FirstDayAsUTC()) != 0));
```
Please do not use <code>GetDateAsInt</code> here. The fact that it returns 0 for time_t(0) is only a coincidence. The semantics of GetDateAsInt is to obtain a unique index value for the GUI spinners, no more no less. I was already warned by a code reviewer about the bad naming of that method. I will change the name soon to reflect it's actual purpose.

What you actually need is <code>CTimerInfoTag::IsFirstDayAtAnyTime</code>, which does not yet exist. But it should be a no brainer for you to implement it, right.

metaron-uk replied:
OK - thanks for the warning - but I'm planning to un-do the 'first day'/'start/end' interactions. They are getting increasingly complex and I agree with you that they will probably break something later. (see my forum post http://forum.kodi.tv/showthread.php?tid=227026&pid=2057604#pid2057604)

ksooo commented:
Regarding first day in the past, look here: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp#L818 Maybe a special handling for first day is needed here?

Regarding strSummary: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/timers/PVRTimerInfoTag.cpp#L336 Maybe also a case for <code>CTimerInfoTag::IsFirstDayAtAnyTime</code> instead of <code>CDateTime::IsValid</code>? I think time_t(0) (means 1/1/1970) unfortunately is valid...

As I said, I cannot really test the first day stuff myself. I can only check that my changes do not introduce regressions (for tvheadend). This is the reason for always asking you whether my changes in this area actually work for you.

metaron-uk commented:
Great! Thanks for the heads-up that should help a lot.

ksooo commented:
@metaron-uk if you rework this PR, please do me a favour and keep it "small", esp. keep PVR add-on API changes out, if possible. 

I say that because I know that you have several ideas in that direction. 

PRs containing API changes tend to be a little "complicated" and therefore take more time until they are merged.

metaron-uk commented:
Sure - I was planning to break any changes into easy to follow chunks - certainly any API changes would come as separate PRs. I spent yesterday debugging the first-day code in my working pvr.mythtv code, which is the cause of the delay :(.

For this PR I'm working on a solution that can cope with days in the past (all types of 'day' currently)
If I find any more bits of code forcing values to be >now(), other than when 1st populating the Timer settings for a new timer, I'll include them in this PR.
Mythtv includes a manual timer type so I'm using that to check I haven't broken the rest of the 'days'.
I also plan to revert 'firstDay' being used to set the date part of the startTime/endTime - leave that to the addon if it wants to / a better solution using a separate API change PR.
Currently firstDay seems to default to startTime if the timer supports (start+end && first day) and firstDay is 0. This seems sensible to me so I won't change that (current mythtv code also relies on it).

When I'm done, do you want me to re-name ```GetDateAsInt``` to ```GetDateAsIndex``` as a separate commit, or keep that for a separate PR? I'm in that bit of the code. (the way I'm headed it won't be a simple int conversion anymore anyhow)

ksooo commented:
@metaron-uk sounds all very good to me. And no rush, take your time. Regarding renaming of GetDateAsInt, yes, please do it as a separate commit in this PR.

metaron-uk committed:
[PVR] Timer settings dialog: support unlisted date

Adds a spinner option equal to the current value
if that value isn't within the normal range

metaron-uk committed:
[PVR] Timer settings dialog: rename SetDateFromInt

Old name considered misleading. No change to functionality.
CGUIDialogPVRTimerSettings::SetDateFromInt -> SetDateFromIndex
CGUIDialogPVRTimerSettings::GetDateAsInt -> GetDateAsIndex

metaron-uk commented:
@ksooo - here you go. Hopefully this counts as a 'small' PR. I'll look at strSummary next.
I've chosen -3months, +1year as the supported range.
pvr.mythtv doesn't have a 'manual repeating' timer, so that still needs testing. 
I stole the condition from the 'conditional visibility' section though, so hopefully it will work.

metaron-uk committed:
[PVR] Timer settings dialog: Add weekday for dates

metaron-uk commented:
@ksooo - OK I've made date range non-static. I've also fixed when you remove midnight from a timer. Both these commits will be squashed when I've got the rest of your comments.

I also had a brainwave and added 'day of the week' to the date spinner in an extra commit :) 
I'd be happy to make this a separate PR if you think it make this one 'too big'.
![screenshot005](https://cloud.githubusercontent.com/assets/12870817/8838117/389bd644-30c2-11e5-85c2-0324f73ae450.png)
![screenshot006](https://cloud.githubusercontent.com/assets/12870817/8838118/389e6e36-30c2-11e5-871c-808a1d51ad86.png)

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -144,6 +147,12 @@ void CGUIDialogPVRTimerSettings::SetTimer(CFileItem *item)
   InitializeChannelsList();
   InitializeTypesList();
 
+  //Initialise the spinner date selection range
```
Please insert a blank between // and the text

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -561,20 +570,64 @@ void CGUIDialogPVRTimerSettings::Save()
-
-    if (m_endLocalTime < m_startLocalTime)
+    if (m_timerType->SupportsStartEndTime() && //i.e. has start/end time field
+        m_timerType->IsRepeating())            //but has no start day / end day spinners
```
Please insert a blank between // and the text; here and elsewhere.

ksooo replied:
Not quite sure whether this condition is correct. What exactly do you want to achieve?

metaron-uk replied:
That the manipulation of startTime and endTime only takes place when there is a start/end clock field but no associated start/end date spinner for the user to interact with.
This is the condition that needs to be confirmed cworking using a manual repeating timer.

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -561,20 +570,64 @@ void CGUIDialogPVRTimerSettings::Save()
+        m_startLocalTime.GetAsSystemTime(timerStartTime);
+        SetTimeFromSystemTime(oldStartLocalTime, timerStartTime);
+        m_timerInfoTag->SetStartFromLocalTime(oldStartLocalTime);
+      }
```
I do not understand the purpose of this else block. If you want to guarantee that m_timerInfoTag's start time is untouched in case value is out of covereded time range, why you just don't write to it? Isn't this the best way to ensure it is unchnaged? Short: Is the else block needed at all?

metaron-uk replied:
If startTime is set to a value outside the covered range for dates (i.e. the < or > list values, or some other value specified by the backend which was set up in the past) I don't want to change it as it will affect the backend startdate in an undefined way which might break the backend rule.

However, clock values could still be acessible to the user, where the full clock range is always accessible, so I still want to be able to update the clock part of the startTime field without touching the date bit - that's what the else case does.

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -561,20 +570,64 @@ void CGUIDialogPVRTimerSettings::Save()
+        SetTimeFromSystemTime(oldEndLocalTime, timerEndTime);
+        m_timerInfoTag->SetEndFromLocalTime(oldEndLocalTime);
+      }
+    }
```
Same here. Is this else block needed?

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -657,15 +712,30 @@ void CGUIDialogPVRTimerSettings::AddCondition(
+  else if (date < m_minDateCovered)
+    (m_minDateCovered - CDateTimeSpan(1, 0, 0, 0)).GetAsTime(t);
+  else
+    date.GetAsTime(t);
```
IMO, no change is needed to this method at all. All you need for the spinner to work is a unique value for every entry. This is achieved if you just use the numerical time_t value. Whether the time value is in covered range is completely unrelated here. Mapping of values out of range is imo unnecessary and makes things only more complicated.

metaron-uk replied:
This allows for the 'no change' < and > list options - which coupled with the 'else' cases above is the clever bit.
It allows us to tolerate (and edit) timers with out of range dates without affecting these dates unless the user specifically selects a value in the spinner's valid date range.

ksooo replied:
I'm not sure whether this is clever or just makes things complicated witout a real value add. See my comment regarding the > and < entries feature as such.

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -657,15 +712,30 @@ void CGUIDialogPVRTimerSettings::AddCondition(
+  if (date < m_minDateCovered)
+    return false;
+  return true;
+}
```
Could by simplified:
<pre>const CDateTime date(datetime.GetYear(), datetime.GetMonth(), datetime.GetDay(), 0, 0, 0);
return (date >= m_minDateCovered) && (date <= m_maxDateCovered) ;</pre>

metaron-uk replied:
Elegant!

ksooo replied:
Reads really "natural", I'd say.

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -813,25 +893,22 @@ void CGUIDialogPVRTimerSettings::DaysFiller(
   {
     list.clear();
     current = 0;
+    CDateTime time = pThis->m_minDateCovered;
```
Prefer init over ctor + assign:
<code>CDateTime time(pThis->m_minDateCovered</code>);</code>

metaron-uk replied:
OK - I'm still a c coder at heart is seems ;)

ksooo commented on xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:
```diff
@@ -813,25 +893,22 @@ void CGUIDialogPVRTimerSettings::DaysFiller(
-      time.GetYear() + 1, time.GetMonth(), time.GetDay() - 1, time.GetHour(), time.GetMinute(), time.GetSecond());
-
-    while (time <= yesterdayPlusOneYear)
+    list.push_back(std::make_pair("< " + time.GetAsLocalizedDate(), pThis->GetDateAsIndex(time - CDateTimeSpan(1, 0, 0, 0))));
```
Wouldn't it be a good idea to add the > and < entries only if start time and|or end time are out of range? What happens if start and|or end time of the timer are not out of range and the user selects one of the two "special" entries? Will the user understand, that "< day x" then actually means "day x - 1"? In the case of one or both values are actually out of range (but heaving a concrete value that is not displayed here), will the user understand that "< day x" actually means some other concrete day (that is not displayed here)? 

The longer I think about the > and < entry idea, the less I tend to like it.

My proposal: Leave this stuff out. It makes the code complicated and users will not really understand the "feature" (at least from my point of view). If a value is out of range, add just one extra entry for exactly this value here, labeled with the concrete out of range day.

metaron-uk replied:
Now that solution (the concrete out of range value) I really like and isn't one I had considered.
I didn't know that would be possible using a 'generic' dates list - however if the range is specific to a particular timer (which it now is I believe) then this would work like a dream. Will do.

ksooo replied:
Glad you like the idea. You're really open-minded. It's a pleasure working with you. Thx for that.

metaron-uk commented:
@ksooo OK all your comments are in (I hope!). I think I've got the hang of spaces after //
Adding an extra option for 'current date' made the second commit much simpler. I like the way it works.

Q: Could one do something similar for an addon supplied list (say if valid priority range =-99:+99 but you only wanted to show +/-5 most of the time to reduce clutter...?)

metaron-uk commented:
OK @ksooo I think I've got your comments in, but for some reason they've diappeared from github, along with my last one related to using long date format :-(
![screenshot007](https://cloud.githubusercontent.com/assets/12870817/8857239/3f740936-3168-11e5-9f61-8a4bced796f1.png)

If you're still not sure about 223e46d as a solution, I'll just drop it from this PR - let me know.

ksooo commented:
Almost ready to go. Could you please do me a last favour and adjust the commit messages to have the form "[PVR] Timer settings dialog: (rest of commit msg)" Or use dialogue, if you prefer ;-)

Thanks again for your patience and cooperation.

metaron-uk commented:
Here you go. Thanks for the constructive editorial input and quick responses :+1:

ksooo commented:
jenkins build this please

ksooo commented:
Jenkins errors are not related to this PR (adsp.* addon build failures).
